### PR TITLE
BAVL-264:  Admin forms, routes, service for CSV data extracts

### PR DESCRIPTION
**Pull request by timharrison-moj**
Contains
* Admin tile on homepage - set by role BVLS_ADMIN
* Admin area
* Tiles / forms / validation to enter dates/days for both options for both court and probation
* Service and ApiClient calls
* Cypress tests

screencapture-localhost-3000-admin-extract-by-hearing-date-2024-08-05-12_47_34

### CODE REVIEW

File: server/data/bookAVideoLinkApiClient.ts [hunk 1]
```diff
@@ -120,4 +121,38 @@ export default class BookAVideoLinkApiClient extends RestClient {
...
+    date: Date,
+    days: number,
+    response: express.Response,
+    user: Express.User,
+  ): Promise<void> {
+    return this.pipeFileStream(
+      {
+        path: `/download-csv/court-data-by-hearing-date`,
+        query: { 'start-date': formatDate(date, 'yyyy-MM-dd'), days: days.toString() },
+        headers: { 'content-type': 'text/csv' },
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply server/data/bookAVideoLinkApiClient.ts:135]: Content type header?  Just something I was playing with.

File: server/routes/interfaces/pageHandler.ts [hunk 2]
```diff
@@ -7,5 +7,6 @@ export interface PageHandler {
   PAGE_NAME: Page
   GET(req: Request, res: Response, next?: NextFunction): Promise<void>
   POST?(req: Request, res: Response, next?: NextFunction): Promise<void>
+  // STREAM?(req: Request, res: Response, next?: NextFunction): Promise<Readable>
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply server/routes/interfaces/pageHandler.ts:10]: Again - one of my past experiments..

[the remark this answers is not in the document]
Comment (timharrison-moj) [reply server/routes/interfaces/pageHandler.ts:10]: Done

File: server/routes/journeys/admin/handlers/extractByHearingDateHandler.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  @Validator(date => date <= startOfToday(), { message: "Enter a date which is before or on today's date" })
+  @IsValidDate({ message: 'Enter a valid start date' })
+  @IsNotEmpty({ message: 'Enter a start date' })
+  startDate: Date
+
+  @Expose()
+  @Transform(({ value }) => +value)
+  @IsInt({ message: 'Enter a whole number between 1 and 365' })
+  @Min(1)
+  @Max(365)
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply server/routes/journeys/admin/handlers/extractByHearingDateHandler.ts:25]: Done

File: server/views/pages/admin/extractByBookingDate.njk [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+                    name: 'startDate',
+                    label: {
+                        text: "Start date",
+                        classes: 'govuk-label--s'
+                    },
+                    hint: {
+                        text: 'Enter the start date, for example ' + exampleDatePickerDate() + ', or click on the calendar to select'
+                    },
+                    errorMessage: validationErrors | findError('startDate'),
+                    value: formResponses.startDate | formatDate('dd/MM/yyyy')
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply server/views/pages/admin/extractByBookingDate.njk:32]: Done

File: server/views/pages/admin/extractByHearingDate.njk [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+                    label: {
+                        text: "Start date",
+                        classes: 'govuk-label--s'
+                    },
+                    hint: {
+                        text: 'Enter the start date, for example ' + exampleDatePickerDate() + ', or click on the calendar to select'
+                    },
+                    classes: 'hmpps-datepicker--fixed-width',
+                    errorMessage: validationErrors | findError('startDate'),
+                    value: formResponses.startDate | formatDate('dd/MM/yyyy')
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply server/views/pages/admin/extractByHearingDate.njk:33]: done.

File: server/views/pages/admin/extractByHearingDate.njk [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+{% block content %}
+    <div class="govuk-grid-row">
+        <div class="govuk-grid-column-three-quarters">
+            <h1 class="govuk-heading-l">{{ pageTitle }}</h1>
+        </div>
+    </div>
+
+    <div class="govuk-grid-row">
+        <div class="govuk-grid-column-three-quarters">
+            <form method='POST'>
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply server/views/pages/admin/extractByHearingDate.njk:18]: Might have to be one for later..

File: server/routes/journeys/admin/handlers/extractByBookingDateHandler.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  @IsNotEmpty({ message: 'Enter a start date' })
+  startDate: Date
+
+  @Expose()
+  @Transform(({ value }) => +value)
+  @IsInt({ message: 'Enter a whole number between 1 and 365' })
+  @Min(1)
+  @Max(365)
+  @IsNotEmpty({ message: 'Enter a number of days from the start date to extract data for' })
+  numberOfDays: number
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply server/routes/journeys/admin/handlers/extractByBookingDateHandler.ts:27]: Done

File: server/data/bookAVideoLinkApiClient.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply server/data/bookAVideoLinkApiClient.ts:135]: Removed.

File: server/routes/journeys/admin/index.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+import AdminHandler from './handlers/adminHandler'
+import { PageHandler } from '../../interfaces/pageHandler'
+import logPageViewMiddleware from '../../../middleware/logPageViewMiddleware'
+import validationMiddleware from '../../../middleware/validationMiddleware'
+import ExtractByHearingDateHandler from './handlers/extractByHearingDateHandler'
+import ExtractByBookingDateHandler from './handlers/extractByBookingDateHandler'
+
+export default function Index({ auditService, videoLinkService }: Services): Router {
+  const router = Router({ mergeParams: true })
+
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply server/routes/journeys/admin/index.ts:13]: Yes.. will do.

File: server/views/pages/admin/admin.njk [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+{% set pageTitle = "Administration" %}
+
+{% block content %}
+    <h1 class="govuk-heading-l govuk-!-margin-top-7">{{ pageTitle }}</h1>
+    <div class="card-group govuk-grid-row">
+        <div class="govuk-grid-column-one-third card-group__item">
+            {{ card({
+                "href": "/admin/extract-by-hearing-date",
+                "clickable": "true",
+                "heading": "Extract data by hearing date",
```
Comment (matt-crowson-moj) [inline server/views/pages/admin/admin.njk:13]: `Extract data by hearing date` or `Extract data by hearing/meeting date` 🤷

**Review by matt-crowson-moj — APPROVED** [review]
LGTM

**Pull request merged**